### PR TITLE
tools: add "set-osdmap" command to ceph-objectstore-tool

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: #12193
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
@dzafman, i am not sure my change is aligned with what we have in ceph-objectstore-tool. or i should fix this issue in anther way? could you take a look? thanks.

File: src/tools/ceph_objectstore_tool.cc [hunk 1]
```diff
@@ -2666,7 +2702,7 @@ int main(int argc, char **argv)
       formatter->flush(cout);
       cout << std::endl;
     } else {
-      cerr << "Must provide --op (info, log, remove, export, import, list, fix-lost, list-pgs, rm-past-intervals)"
+      cerr << "Must provide --op (info, log, remove, export, import, list, fix-lost, list-pgs, rm-past-intervals, set-osdmap)"
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: Missed an equivalent change above at line 1870.

**Comment by dzafman** [comment]
How about adding a get-osdmap feature?  What about the incrementals?

Would you add a simple test case to test/ceph_objectstore_tool.py?

File: src/tools/ceph_objectstore_tool.cc (hunk 1 above)
Comment (tchaikov) [inline src/tools/ceph_objectstore_tool.cc]: thanks. will do

**Comment by tchaikov** [comment]
> How about adding a get-osdmap feature? What about the incrementals?

will do. i am sure about the incrementals, maybe we should simply **remove** it, in case OSD will share it with others? otherwise we need to prepare an incremental between the epoch we are overwriting and the previous one. OSD will ignore the not found incremental when sharing the osdmap with its peers, and fall back to the full map.

> Would you add a simple test case to test/ceph_objectstore_tool.py?

will do.

**Comment by tchaikov** [comment]
@dzafman, thanks for your review =). will update this pull request later on.

**Comment by tchaikov** [comment]
> i am not sure about the incrementals, maybe we should simply remove it, in case OSD will share it with others? otherwise we need to prepare an incremental between the epoch we are overwriting and the previous one. OSD will ignore the not found incremental when sharing the osdmap with its peers, and fall back to the full map.

per the discussion today, instead of removing the incrementals, we should rewrite them using the ones extracted from the monitor. see #5052 .

File: src/tools/ceph_objectstore_tool.cc [hunk 2]
```diff
@@ -1973,7 +2010,7 @@ int main(int argc, char **argv)
   ObjectStoreTool tool = ObjectStoreTool(file_fd, dry_run);
 
   if (vm.count("file") && file_fd == fd_none && !dry_run) {
-    cerr << "--file option only applies to import or export" << std::endl;
+    cerr << "--file option only applies to import, export or set-osdmap" << std::endl;
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: get-osdmap is included here.

**Comment by dzafman** [comment]
@tchaikov You'll need rebase this now that #5197 merged.

**Comment by tchaikov** [comment]
@dzafman, updated, rebased and re-pushed.

File: src/tools/ceph_objectstore_tool.cc [hunk 3]
```diff
@@ -1852,9 +1921,12 @@ int main(int argc, char **argv)
     ("pgid", po::value<string>(&pgidstr),
      "PG id, mandatory for info, log, remove, export, rm-past-intervals")
     ("op", po::value<string>(&op),
-     "Arg is one of [info, log, remove, export, import, list, fix-lost, list-pgs, rm-past-intervals, set-allow-sharded-objects, dump-journal, dump-super, meta-list]")
+     "Arg is one of [info, log, remove, export, import, list, fix-lost, list-pgs, rm-past-intervals, set-allow-sharded-objects, dump-journal, dump-super, meta-list, "
+	 "get-osdmap, set-osdmap, get-inc-osdmap, set-inc-osdmap]")
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: There is an equivalent error message to this usage at line 2571 that needs the new ops listed.

File: src/tools/ceph_objectstore_tool.cc [hunk 4]
```diff
@@ -719,6 +719,74 @@ int ObjectStoreTool::export_files(ObjectStore *store, coll_t coll)
...
+  OSDMap::Incremental inc;
+  bufferlist::iterator it = bl.begin();
+  inc.decode(it);
+  if (e == 0) {
+    e = inc.epoch;
+  } else if (e != inc.epoch) {
+    cerr << "incremental.epoch mismatch: "
+	 << inc.epoch << " != " << e << ".\n"
+	 << "But will continue anyway." << std::endl;
+  }
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: I think you should set inc.epoch = e here.  Also, maybe you want to utilize the existing --force flag to allow this but only if the --force set.

File: src/tools/ceph_objectstore_tool.cc [hunk 5]
```diff
@@ -719,6 +719,74 @@ int ObjectStoreTool::export_files(ObjectStore *store, coll_t coll)
...
+  const ghobject_t inc_oid = OSD::get_inc_osdmap_pobject_name(e);
+  if (!store->exists(coll_t::meta(), inc_oid)) {
+    cerr << "inc-osdmap (" << inc_oid << ") does not exist, "
+	 << "I am supposed to overwrite it." << std::endl;
+    return -ENOENT;
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: Should you be able to create a new incremental if --force is set?

File: src/tools/ceph_objectstore_tool.cc [hunk 6]
```diff
@@ -719,6 +719,74 @@ int ObjectStoreTool::export_files(ObjectStore *store, coll_t coll)
...
+int set_osdmap(ObjectStore *store, bufferlist& bl) {
+  OSDMap osdmap;
+  osdmap.decode(bl);
+  if (debug)
+    cerr << osdmap << std::endl;
+  const ghobject_t full_oid = OSD::get_osdmap_pobject_name(osdmap.get_epoch());
+  if (!store->exists(coll_t::meta(), full_oid)) {
+    cerr << "osdmap (" << full_oid << ") does not exist, "
+	 << "I am supposed to overwrite it." << std::endl;
+    return -ENOENT;
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: Same with --force

File: src/tools/ceph_objectstore_tool.cc (hunk 4 above)
Comment (tchaikov) [inline src/tools/ceph_objectstore_tool.cc]: > I think you should set inc.epoch = e here. 

this is intentional to not avoid changing the injected osdmap. this helps to simplify the test, where we

1. read an inc_map from e2
1. put inc_map.e2 to e1
1. read inc_map.e1 from e1
1. compare inc_map.e1 and inc_map.e2

if we set the inc.epoch. we will need to do the test in a more complicated way.

> maybe you want to utilize the existing --force flag to allow this but only if the --force set.

thanks, will do.

File: src/tools/ceph_objectstore_tool.cc (hunk 5 above)
Comment (tchaikov) [inline src/tools/ceph_objectstore_tool.cc]: will do.

**Comment by tchaikov** [comment]
test failed, /me looking

**Comment by tchaikov** [comment]
test failed, /me looking

**Comment by tchaikov** [comment]
test failed, /me looking

**Comment by tchaikov** [comment]
test failed, /me looking

**Pull request merged**